core-bundle/tests/TestCase.php

```diff
@@ -15,13 +15,95 @@
 use Contao\CoreBundle\Routing\ScopeMatcher;
 use Contao\CoreBundle\Session\Attribute\ArrayAttributeBag;
 use Contao\TestCase\ContaoTestCase;
+use Roave\BetterReflection\BetterReflection;
 use Symfony\Component\HttpFoundation\RequestMatcher;
 use Symfony\Component\HttpFoundation\Session\Session;
 use Symfony\Component\HttpFoundation\Session\SessionInterface;
 use Symfony\Component\HttpFoundation\Session\Storage\MockArraySessionStorage;
 
 abstract class TestCase extends ContaoTestCase
 {
+    private static array $betterReflectionCache = [];
+
+    private array $backupServerEnvGetPost = [];
+
+    public static function setUpBeforeClass(): void
+    {
+        parent::setUpBeforeClass();
+
+        if (!\defined('TL_FILES_URL')) {
+            \define('TL_FILES_URL', '');
+        }
+    }
+
+    protected function tearDown(): void
+    {
+        unset($GLOBALS['TL_CONFIG']);
+
+        parent::tearDown();
+    }
+
+    protected function backupServerEnvGetPost(): void
+    {
+        $this->backupServerEnvGetPost = [
+            '_SERVER' => $_SERVER,
+            '_ENV' => $_ENV,
+            '_GET' => $_GET,
+            '_POST' => $_POST,
+        ];
+    }
+
+    protected function restoreServerEnvGetPost(): void
+    {
+        $_SERVER = $this->backupServerEnvGetPost['_SERVER'] ?? $_SERVER;
+        $_ENV = $this->backupServerEnvGetPost['_ENV'] ?? $_ENV;
+        $_GET = $this->backupServerEnvGetPost['_GET'] ?? [];
+        $_POST = $this->backupServerEnvGetPost['_POST'] ?? [];
+    }
+
+    /**
+     * @param array<int, class-string|array{0: class-string, 1: array<int, string>}> $classNames
+     */
+    protected function resetStaticProperties(array $classNames): void
```

Comment:
This looks suspicious to me. Most of these classes have a `reset` method that is called by the `ContaoFramework`. We should call that method instead, and add `reset` methods if some are missing (e.g. the `System` class) to make sure it works consistently.

Reply:
As this is only used in the unit tests, using reflection should be OK here IMO.
Also, several tests need to reset third party classes like `Request`, `Terminal`, `Process` and so on.

But we could implement a `method_exists($class, 'reset')` and call `$class::reset()` if it exists here, WDYT?

Reply:
By the way: I tried this in #2518, but it got a convoluted complicated thing and probably not worth for the legacy code.  :see_no_evil: While I generally agree with @aschempp, I think it's probably most helpful adjusting it in the tests with reflection here. :+1:

Reply:
> But we could implement a method_exists($class, 'reset') and call $class::reset() if it exists here, WDYT?

It should be sufficient to call `ContaoFramework#reset()` but unfortunately this is a class method.

Reply:
How would we reset the three properties from here https://github.com/symfony/symfony/blob/44db49fb8f6d70c4341120a3f8f1b1a1a0dfa014/src/Symfony/Component/Console/Terminal.php#L16-L18 ?

Reply:
> But we could implement a `method_exists($class, 'reset')` and call `$class::reset()` if it exists

Implemented in 042e1a647d543a33efbeb3a001348f2528b718ba